src/test/java/org/springframework/integration/aws/outbound/KplMessageHandlerTests.java

```diff
@@ -0,0 +1,170 @@
+		verify(this.kinesisProducer, Mockito.times(0)).getOutstandingRecordsCount();
+		UserRecord userRecord = userRecordRequestArgumentCaptor.getValue();
+		assertThat(userRecord.getStreamName()).isEqualTo("foo");
+		assertThat(userRecord.getPartitionKey()).isEqualTo("fooKey");
+		assertThat(userRecord.getExplicitHashKey()).isNull();
+	}
+
+	@Test
+	@SuppressWarnings("unchecked")
+	void testKPLMessageHandler_raw_payload_backpressure_capacity_available_test() {
+		given(this.kinesisProducer.addUserRecord(any(UserRecord.class)))
+				.willReturn(mock());
+		this.kplMessageHandler.setBackPressureThreshold(2);
+		given(this.kinesisProducer.getOutstandingRecordsCount())
+				.willReturn(1);
+		final Message<?> message = MessageBuilder
+				.withPayload("message1")
+				.setHeader(AwsHeaders.PARTITION_KEY, "fooKey")
+				.setHeader(AwsHeaders.SEQUENCE_NUMBER, "10")
+				.setHeader("foo", "bar")
+				.build();
+
+
+		ArgumentCaptor<UserRecord> userRecordRequestArgumentCaptor = ArgumentCaptor
+				.forClass(UserRecord.class);
+
+		this.kinesisSendChannel.send(message);
+		verify(this.kinesisProducer).addUserRecord(userRecordRequestArgumentCaptor.capture());
+		verify(this.kinesisProducer, Mockito.times(1)).getOutstandingRecordsCount();
+		UserRecord userRecord = userRecordRequestArgumentCaptor.getValue();
+		assertThat(userRecord.getStreamName()).isEqualTo("foo");
+		assertThat(userRecord.getPartitionKey()).isEqualTo("fooKey");
+		assertThat(userRecord.getExplicitHashKey()).isNull();
+	}
+
+	@Test
+	@SuppressWarnings("unchecked")
+	void testKPLMessageHandler_raw_payload_backpressure_insufficient_capacity_test() {
+		given(this.kinesisProducer.addUserRecord(any(UserRecord.class)))
+				.willReturn(mock());
+		this.kplMessageHandler.setBackPressureThreshold(2);
+		given(this.kinesisProducer.getOutstandingRecordsCount())
+				.willReturn(5);
+		final Message<?> message = MessageBuilder
+				.withPayload("message1")
+				.setHeader(AwsHeaders.PARTITION_KEY, "fooKey")
+				.setHeader(AwsHeaders.SEQUENCE_NUMBER, "10")
+				.setHeader("foo", "bar")
+				.build();
+
+		try {
+			this.kinesisSendChannel.send(message);
+		}
+		catch (Exception ex) {
+			assertThat(ex).isNotNull();
+			assertThat(ex.getCause()).isNotNull();
+			assertThat(ex.getCause().getClass()).isEqualTo(KPLBackpressureException.class);
+			assertThat(ex.getCause().getMessage()).isEqualTo("Buffer already at max capacity.");
+		}
+
+		verify(this.kinesisProducer, Mockito.times(0)).addUserRecord(any(UserRecord.class));
+		verify(this.kinesisProducer, Mockito.times(1)).getOutstandingRecordsCount();
+	}
+
+	@AfterEach
+	public void tearDown() {
+		clearInvocations(this.kinesisProducer);
+	}
+
+	@Configuration
+	@EnableIntegration
+	public static class ContextConfiguration {
+
+		@Bean
+		public KinesisProducer kinesisProducer() {
+			return mock();
+		}
+
+		@Bean
+		@ServiceActivator(inputChannel = "kinesisSendChannel")
```

Comment:
Any chances to have some test where we indeed use one of our advice , e.g. the mentioned retry one?